### PR TITLE
ISPN-14863 Implement RESP "cluster nodes" command

### PR DESCRIPTION
**Pull request by jabolina**
https://issues.redhat.com/browse/ISPN-14963

### CODE REVIEW

File: server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/NODES.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+                  response.append(topology.getTopologyId()).append(' ');
+                  response.append(health).append(' ');
+                  serializeSegments(response, owner);
+               }
+               return "$" + response.length() + CRLF + response + CRLF;
+            });
+   }
+
+   private static void serializeSegments(StringBuilder response, IntSet ranges) {
+      BitSet bitSet = BitSet.valueOf(ranges.toBitSet());
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/NODES.java:156]: This should be able to just use the ranges.iterator instead. But this change is not required as it is a pretty small optimization I suspect.

Comment (wburns) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/NODES.java:156]: Another option is adding nextSetBit method to IntSet which sounds like a good idea to me. It should be pretty trivial to add the method to the various implementations.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/NODES.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+      CacheManagerInfo info = ecm.getCacheManagerInfo();
+      String name = info.getNodeName();
+      Address address = findPhysicalAddress(ecm);
+      int port = findPort(address);
+      String addressString = address != null ? getOnlyIp(address) : info.getNodeAddress();
+
+      StringBuilder sb = new StringBuilder();
+      sb.append(name).append(' ');
+      sb.append(addressString).append(':').append(port).append('@').append(port);
+      sb.append(",,shard-id=").append(name).append(' ');
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/NODES.java:192]: Is the space at the end always needed?

**Review by wburns — COMMENTED** [review]
LGTM, just some minor comments. I also am not familiar with the actual REDIS output, but I assume you double checked the validity.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/NODES.java (hunk 2 above)
Comment (jabolina) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/NODES.java:192]: Yeah, in this format, they use space as a separator.

Comment (wburns) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/NODES.java:192]: Okay, I was thinking it was a bit of a waste if it was always there, even for the last node. But if that is how they want it :)

File: server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/NODES.java (hunk 1 above)
Comment (jabolina) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/NODES.java]: Added the `nextSetBit` method, and updated here and the SHARDS command.

**Comment by jabolina** [comment]
Updated with suggestions.

**Comment by jabolina** [comment]
I need to find another way to check this beyond using the Lettuce client to validate the response.

**Comment by jabolina** [comment]
Doing a comparison by running with `echo 'cluster-enabled yes' | redis-server -`, my server is version Redis 7.0.10 (00000000/0) 64 bit and it is assigned a few slots automatically. Running the command with `redis-cli`:

```text
127.0.0.1:6379> CLUSTER NODES
308345ab50e60cdf538e30d2fd5e3de671abfb0c :6379@16379 myself,master - 0 0 0 connected 238 5282 5798 9755 12291 12539
127.0.0.1:6379>
```

Running with ISPN:

```text
127.0.0.1:11222> CLUSTER NODES
jbolina-64043 172.17.0.1:7800@7800,,shard-id=jbolina-64043 myself,master - 0 1686081424 1 connected 0-255
127.0.0.1:11222>
```

The ISPN output includes the `shard-id` in the auxiliary field. Per the doc, this field is always included in version >= 7.2.0.
I need to update to return the correct address in the protocol server.

**Comment by jabolina** [comment]
Doing a test with three servers (server-1, server-2, server-3):

```text
127.0.0.1:11222> cluster nodes
server-1 127.0.0.1:11222@60836,,shard-id=server-1 myself,master - 0 1686141297 9 connected 0-255
server-2 127.0.0.1:11232@60836,,shard-id=server-2 master - 0 1686141297 9 connected 0-255
server-3 127.0.0.1:11322@60836,,shard-id=server-3 master - 0 1686141297 9 connected 0-255
127.0.0.1:11222>
```

I think that should be it, going to squash the commits. I guess a point of attention is the way I did to retrieve the resp server.

**Comment by tristantarrant** [comment]
> I need to find another way to check this beyond using the Lettuce client to validate the response.

Maybe introduce a Vert.x Redis client test ?

**Comment by jabolina** [comment]
I think adding the Vert.x might be worth it, especially since we're missing more throughout coverage in cluster mode, checking the CRC16 implementation. Although, the client seems to use the `CLUSTER SLOTS` command ([link](https://github.com/vert-x3/vertx-redis-client/blob/224eb8af8d063ec8152caff75a49f900ba3c0c19/src/main/java/io/vertx/redis/client/impl/RedisClusterClient.java#L227)). 

The `CLUSTER SLOTS` is next on my list. I'll add the client then.

**Comment by jabolina** [comment]
Pushed a small change. I was not retrieving the segments owned (that's why all nodes had 0-255). Now we have:

```text
jbolina-22538 127.0.0.1:11222@54392,,shard-id=jbolina-22538 myself,master - 0 1686572258 5 connected 74-136 142-145 157-168 181-181 187-210 212-217 219-219 225-226 230-247
jbolina-61882 127.0.0.1:11322@54392,,shard-id=jbolina-61882 master - 0 1686572258 5 connected 0-73 137-141 146-156 169-180 182-186 211-211 218-218 220-224 227-229 248-255
```

**Comment by karesti** [comment]
@tristantarrant @wburns something else ?

File: server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/NODES.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+      }
+
+      CompletionStage<CharSequence> cs;
+      synchronized (this) {
+         if (!currentCH.equals(hash)) {
+            requestPerOwner.clear();
+            hash = currentCH;
+         }
+         EmbeddedCacheManager ecm = SecurityActions.getEmbeddedCacheManager(respCache);
+         cs = requestPerOwner.computeIfAbsent(ecm.getAddress(), ignore -> requestClusterInformation(handler, ctx, ecm, topology));
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/NODES.java:90]: Isn't `ecm.getAddress()` always the same thing irrespective of the consistent hash? Not sure I understand the requestPerOwner map usage.

Comment (jabolina) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/NODES.java:90]: Not always, it depends on the server handling the request. The result is slightly different and includes a `myself` in the returned nodes list. We have a test covering this, too. With two servers, each client connects to one server.

Comment (jabolina) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/NODES.java]: Sorry, totally missed the point. Updated now.

**Comment by wburns** [comment]
Integrated into main, thanks @jabolina !

**Pull request merged**